### PR TITLE
[CAMEL-14227]add a new camel-rest-openapi component and use apicurio-…

### PR DESCRIPTION
**Pull request by ffang**
…data-module which can support OpenApi 2 & 3 at the same time

### CODE REVIEW

File: components/camel-rest-openapi/src/main/docs/rest-openapi-component.adoc [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+
+
+*Since Camel 3.1*
+
+// HEADER START
+*Only producer is supported*
+// HEADER END
+
+The REST OpenApi* configures rest producers from
+http://swagger.io/[OpenApi] (Open API) specification document and
```
Comment (zregvart) [inline components/camel-rest-openapi/src/main/docs/rest-openapi-component.adoc]: We could link to https://www.openapis.org/

File: components/camel-rest-openapi/src/main/docs/rest-openapi-component.adoc [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+// spring-boot-auto-configure options: START
+== Spring Boot Auto-Configuration
+
+When using Spring Boot make sure to use the following Maven dependency to have support for auto configuration:
+
+[source,xml]
+----
+<dependency>
+  <groupId>org.apache.camel</groupId>
+  <artifactId>camel-rest-swagger-starter</artifactId>
```
Comment (zregvart) [inline components/camel-rest-openapi/src/main/docs/rest-openapi-component.adoc]:
```suggestion
  <artifactId>camel-rest-openapi-starter</artifactId>
```

File: components/camel-rest-openapi/src/main/docs/rest-openapi-component.adoc [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+
+
+The component supports 10 options, which are listed below.
+
+
+
+[width="100%",cols="2,5,^1,2",options="header"]
+|===
+| Name | Description | Default | Type
+| *camel.component.rest-swagger.base-path* | API basePath, for example /v2. Default is unset, if set overrides the value present in OpenApi specification. |  | String
```
Comment (zregvart) [inline components/camel-rest-openapi/src/main/docs/rest-openapi-component.adoc]:
```suggestion
| *camel.component.rest-openapi.base-path* | API basePath, for example /v2. Default is unset, if set overrides the value present in OpenApi specification. |  | String
```

I think running a full build would update these tables

File: components/camel-rest-openapi/src/main/docs/rest-openapi-component.adoc [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+| *camel.component.rest-swagger.specification-uri* | Path to the OpenApi specification file. The scheme, host base path are taken from this specification, but these can be overridden with properties on the component or endpoint level. If not given the component tries to load swagger.json resource. Note that the host defined on the component and endpoint of this Component should contain the scheme, hostname and optionally the port in the URI syntax (i.e. \https://api.example.com:8080). Can be overridden in endpoint configuration. |  | URI
+| *camel.component.rest-swagger.ssl-context-parameters* | Customize TLS parameters used by the component. If not set defaults to the TLS parameters set in the Camel context. The option is a org.apache.camel.support.jsse.SSLContextParameters type. |  | String
+| *camel.component.rest-swagger.use-global-ssl-context-parameters* | Enable usage of global SSL context parameters. | false | Boolean
+|===
+// spring-boot-auto-configure options: END
+
+
+== Example: PetStore
+
+Checkout the example in the `camel-example-rest-swagger` project in
```
Comment (zregvart) [inline components/camel-rest-openapi/src/main/docs/rest-openapi-component.adoc]:
```suggestion
Checkout the example in the `camel-example-rest-openapi` project in
```

File: components/camel-rest-openapi/src/main/docs/rest-openapi-component.adoc [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+[source,xml]
+----
+<dependency>
+  <groupId>org.apache.camel</groupId>
+  <artifactId>camel-undertow-starter</artifactId>
+</dependency>
+
+<dependency>
+  <groupId>org.apache.camel</groupId>
+  <artifactId>camel-rest-swagger-starter</artifactId>
```
Comment (zregvart) [inline components/camel-rest-openapi/src/main/docs/rest-openapi-component.adoc]:
```suggestion
  <artifactId>camel-rest-openapi-starter</artifactId>
```

File: components/camel-rest-openapi/src/main/docs/rest-openapi-component.adoc [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+----
+
+Start by defining the _Undertow_ component and the
+_RestOpenApiComponent_:
+
+[source,java]
+----
+@Bean
+public Component petstore(CamelContext camelContext, UndertowComponent undertow) {
+    RestOpenApiComponent petstore = new RestSwaggerComponent(camelContext);
```
Comment (zregvart) [inline components/camel-rest-openapi/src/main/docs/rest-openapi-component.adoc]:
```suggestion
    RestOpenApiComponent petstore = new RestOpenApiComponent(camelContext);
```

File: components/camel-rest-openapi/src/main/java/org/apache/camel/component/rest/openapi/RestOpenApiComponent.java [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+ * clients, but it configures them from OpenApi specification. Client needs to
+ * point to operation that it wants to invoke via REST, provide any additional
+ * HTTP headers as headers in the Camel message, and any payload as the body of
+ * the incoming message.
+ * <p>
+ * Example usage using Java DSL:
+ * <p>
+ *
+ * <pre>
+ * from(...).to("rest-openapi:http://petstore.swagger.io/v2/swagger.json#getPetById")
```
Comment (zregvart) [inline components/camel-rest-openapi/src/main/java/org/apache/camel/component/rest/openapi/RestOpenApiComponent.java]: We could showcase support for OpenAPI 3 here:

```suggestion
 * from(...).to("rest-openapi:https://petstore3.swagger.io/api/v3/openapi.json#getPetById")
```

File: components/camel-rest-openapi/src/main/java/org/apache/camel/component/rest/openapi/RestOpenApiComponent.java [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+ * component in Java DSL:
+ * <p>
+ *
+ * <pre>
+ * Component undertow = new UndertowComponent();
+ * undertow.setSslContextParameters(...);
+ * //...
+ * camelContext.addComponent("myUndertow", undertow);
+ *
+ * from(...).to("rest-openapi:http://petstore.swagger.io/v2/swagger.json#getPetById?componentName=myUndertow")
```
Comment (zregvart) [inline components/camel-rest-openapi/src/main/java/org/apache/camel/component/rest/openapi/RestOpenApiComponent.java]: Perhaps showing 3.0 example?

```suggestion
 * from(...).to("rest-openapi:https://petstore3.swagger.io/api/v3/openapi.json#getPetById?componentName=myUndertow")
```

File: components/camel-rest-openapi/src/main/java/org/apache/camel/component/rest/openapi/RestOpenApiComponent.java [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+ * </pre>
+ *
+ * The most concise way of using this component would be to define it in the
+ * Camel context under a meaningful name, for example:
+ *
+ * <pre>
+ * Component petstore = new RestOpenApiComponent();
+ * petstore.setSpecificationUri("http://petstore.openapi.io/v2/swagger.json");
```
Comment (zregvart) [inline components/camel-rest-openapi/src/main/java/org/apache/camel/component/rest/openapi/RestOpenApiComponent.java]: Also showing 3.0 example

```suggestion
 * petstore.setSpecificationUri("https://petstore3.swagger.io/api/v3/openapi.json");
```

File: platforms/spring-boot/components-starter/camel-rest-openapi-starter/src/main/resources/META-INF/LICENSE.txt [hunk 10]
```diff
@@ -0,0 +1,203 @@
+
```
Comment (zregvart) [inline platforms/spring-boot/components-starter/camel-rest-openapi-starter/src/main/resources/META-INF/LICENSE.txt]: Do we need this, isn't this automatically added?

File: platforms/spring-boot/components-starter/camel-rest-openapi-starter/src/main/resources/META-INF/NOTICE.txt [hunk 11]
```diff
@@ -0,0 +1,11 @@
+   =========================================================================
```
Comment (zregvart) [inline platforms/spring-boot/components-starter/camel-rest-openapi-starter/src/main/resources/META-INF/NOTICE.txt]: Not sure we need this

**Comment by davsclaus** [comment]
I think a new camel-rest-openapi component is better, and then we can deprecate the old swagger. Its still early days in Camel 3 so it's fine.

File: components/camel-rest-openapi/src/test/resources/log4j2.properties [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+## Unless required by applicable law or agreed to in writing, software
+## distributed under the License is distributed on an "AS IS" BASIS,
+## WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+## See the License for the specific language governing permissions and
+## limitations under the License.
+## ---------------------------------------------------------------------------
+
+appender.file.type = File
+appender.file.name = file
+appender.file.fileName = target/camel-rest-swagger-test.log
```
Comment (davsclaus) [inline components/camel-rest-openapi/src/test/resources/log4j2.properties]: should be openapi

File: components/camel-rest-openapi/pom.xml [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+        <version>3.1.0-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>camel-rest-openapi</artifactId>
+
+    <name>Camel :: REST OpenApi</name>
+    <description>Camel REST support using OpenApi</description>
+
+    <properties>
+        <firstVersion>2.19.0</firstVersion>
```
Comment (davsclaus) [inline components/camel-rest-openapi/pom.xml]: Should be 3.1.0

File: platforms/spring-boot/components-starter/camel-rest-openapi-starter/src/main/resources/META-INF/NOTICE.txt (hunk 11 above)
Comment (ffang) [reply platforms/spring-boot/components-starter/camel-rest-openapi-starter/src/main/resources/META-INF/NOTICE.txt]: Both LICENSE.txt and NOTICE.txt are automatically generated and added into the git.

I checked other components starts, they also have those files in git, so I'd keep this new component same as others.

File: components/camel-rest-openapi/src/test/resources/log4j2.properties (hunk 12 above)
Comment (ffang) [reply components/camel-rest-openapi/src/test/resources/log4j2.properties]: fixed accordingly, thanks!

File: components/camel-rest-openapi/src/main/docs/rest-openapi-component.adoc (hunk 1 above)
Comment (ffang) [reply components/camel-rest-openapi/src/main/docs/rest-openapi-component.adoc]: fixed accordingly, thanks!

**Comment by ffang** [comment]
Thanks @zregvart  and @davsclaus  for the through feedback!

And yes, I tend to keep camel-rest-swagger component for a while and deprecate it late on. All others revised accordingly.

**Pull request merged**